File: Week 1 C/credit.c
#include <cs50.h>
#include <stdio.h>
int main(void)
{
    // 计算校验值
    int size = 0;
    int digit;
    int checksum = 0;
    int first;
    int second;
    long number = get_long("Number: ");
    for (int i = 0; number > 0; i++)
    {
        digit = number % 10;
        if (i % 2 == 0)
        {
            checksum += digit;
        }
        else
        {
            if (digit * 2 < 10)
            {
                checksum += digit * 2;
            }
            else
            {
                checksum += 1 + digit * 2 % 10;
            }
        }
        number /= 10;
        second = first;
        first = digit;
        size++;
    }
    // 检测校验值
    if (checksum % 10 == 0)
    {
        // AMEX
        if (size == 15 && first == 3 && (second == 4 || second == 7))
        {
            printf("AMEX\n");
        }
        // MasterCard
        else if (size == 16 && first == 5 && second < 6 && second > 0)
        {
            printf("MASTERCARD\n");
        }
        // VISA
        else if ((size == 13 || size == 16) && first == 4)
        {
            printf("VISA\n");
        }
        // Invalid
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Week 6 Python/credit.py
from cs50 import get_string

n = get_string("Number: ")

sum = 0

"""
for i in range(len(n)):
    i += 1
    digit = int(n[len(n) - 1 - i]) * 2
    if digit > 9:
        digit = 1 + digit - 10
    sum += digit

for i in range(len(n)):
    sum += int(n[len(n) - 1 - i])
    i += 1
"""
# Think why it doesn't work?

for i in range(len(n)):
    digit = int(n[len(n) - 1 - i])
    if i % 2 == 1:
        digit *= 2
        if digit > 9:
            digit = 1 + digit - 10
        sum += digit
    else:
        sum += int(n[len(n) - 1 - i])

if sum % 10 == 0:
    if n[0] == "5" and (n[1] == "1" or n[1] == "2" or n[1] == "3" or n[1] == "4" or n[1] == "5") and len(n) == 16:
        print("MASTERCARD")
    elif n[0] == "4" and (len(n) == 13 or len(n) == 16):
        print("VISA")
    elif n[0] == "3" and (n[1] == "4" or n[1] == "7") and len(n) == 15:
        print("AMEX")
    else:
        print("INVALID")
else:
    print("INVALID")
